### PR TITLE
Added the histogram equalization operator

### LINKED ISSUE #41: Add Histogram Equalization Operator
**Difficulty:** Medium
**Area:** backend, frontend

## Description

Histogram equalization is a technique for improving the contrast of an image by redistributing the intensity values so that they span the full available range. This is particularly useful for images that appear washed out or have poor contrast due to lighting conditions. It is a foundational image processing operation that belongs in any comprehensive image processing toolkit.

OpenCV's `cv2.equalizeHist()` operates on single-channel (grayscale) images. When the input is a color image, the operator must first convert it to grayscale before applying equalization. This is a parameterless operator -- the user simply adds it to the pipeline and it enhances contrast automatically.

## Acceptance Criteria

- [ ] A histogram equalization operator is implemented in the backend following the existing operator pattern
- [ ] The operator handles both grayscale and color input images and outputs a pipeline-compatible image
- [ ] No parameters are required (the operator applies equalization automatically)
- [ ] The operator is registered and available in the processing pipeline
- [ ] A Blockly block exists with no parameter fields, appearing in the conversions category
- [ ] Visual verification: applying the operator to a low-contrast image produces noticeably improved contrast

## Resources

- [OpenCV Histogram Equalization tutorial](https://docs.opencv.org/4.x/d5/daf/tutorial_py_histogram_equalization.html)
- [cv2.equalizeHist API reference](https://docs.opencv.org/4.x/d6/dc7/group__imgproc__hist.html#ga7e54091f0c937d49bf84152a16f76c6e)
- Existing conversion operators in `imagelab-backend/app/operators/conversions/` for patterns

### PR DESCRIPTION
**Pull request by AbhaBarge**
## Description

Implemented the **Histogram Equalization operator for both backend and frontend**.

On the backend, a new HistogramEqualization operator was added under the conversions module, following the existing BaseOperator pattern. The operator applies cv2.equalizeHist() for grayscale images and, for color images, performs equalization on the luminance (Y) channel in YCrCb space to preserve color fidelity. It requires no parameters and has been registered in registry.py to integrate with the processing pipeline.

On the frontend, a corresponding Blockly block was added under the Conversions category with no parameter fields, fully integrated into the existing pipeline flow.

The implementation has been validated using standard and low-contrast grayscale and color images, confirming improved contrast and compatibility with downstream operators.

Fixes #41 

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## How Has This Been Tested?

Describe the tests you ran to verify your changes.
- [x] Existing tests pass
- [ ] New tests added
- [x] Manual testing

## Screenshots (if applicable)

Screenshot 2026-02-26 at 11 05 39 AM
Screenshot 2026-02-26 at 11 07 38 AM
Screenshot 2026-02-26 at 11 15 14 AM
Screenshot 2026-02-26 at 11 23 35 AM

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [x] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [x] Tests pass locally

### CODE REVIEW

**Review by ivantha — APPROVED** [review]
Lgtm 👍🏼 Please rebase the PR to merge.

**Comment by ivantha** [comment]
@AbhaBarge, looks like we have conflicts due to the earlier PRs that were merged. Could you please rebase once again? Thanks

**Comment by ivantha** [comment]
I am still seeing conflicts.

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR adds a Histogram Equalization operator to the backend (Python/OpenCV) and a corresponding Blockly block on the frontend. The backend implementation is mostly sound and follows existing patterns, but contains a potential runtime crash for non-uint8 images and silent failure for unsupported channel counts. More critically, **the frontend changes contain multiple syntax errors** — a missing comma in `categories.ts` and duplicate `tooltip` properties (plus a missing comma) in `conversions.blocks.ts` — that will prevent the frontend from compiling. These bugs appear to stem from a bad cut/paste operation when inserting the new block definition. No automated tests were added for the new operator.

File: imagelab-frontend/src/blocks/definitions/conversions.blocks.ts [hunk 1]
```diff
@@ -86,6 +86,15 @@ export const conversionsBlocks = [
...
+  },
+  {
+    type: "imageconvertions_histogramequalization",
+    message0: "Enhance contrast using histogram equalization",
+    previousStatement: null,
+    nextStatement: null,
+    style: "conversions_style",
+    tooltip: "Automatically improves image contrast using histogram equalization"
     tooltip: "Convert colored (RGB) image to binary with adjustable threshold - Applies a binary threshold to a colored image, converting it to black and white. You can choose between 'Threshold Binary' (pixels above the threshold become white) and 'Threshold Binary Inv' (pixels above the threshold become black). Adjust the threshold value to control which pixels are considered foreground (white) or background (black), and set the max value for the output binary image.",
   },
```
Comment (ivantha) [inline imagelab-frontend/src/blocks/definitions/conversions.blocks.ts]: **[CRITICAL]** Duplicate `tooltip` property and missing comma in histogram equalization block definition

The new `imageconvertions_histogramequalization` block object contains **two `tooltip` properties**, which is invalid. This appears to be a cut/paste error: the long tooltip string belonging to the `ColorToBinary` block (`"Convert colored (RGB) image to binary with adjustable threshold - Applies a binary threshold..."`) was not removed from its original location and now appears as a second, orphaned `tooltip` key inside the new histogram equalization block.

Additionally, the first `tooltip` line is **missing a trailing comma**, creating a secondary syntax error.

In TypeScript, duplicate object keys are a compiler error (`error TS2300`), and the missing comma is a parse error. The net effect is:
1. **Build fails** due to both syntax errors.
2. Even if the errors were silently ignored (e.g., in loose JS), the histogram equalization block would display the wrong tooltip — the `ColorToBinary` long description.

```ts
// Current (broken)
{
  type: "imageconvertions_histogramequalization",
  ...
  tooltip: "Automatically improves image contrast using histogram equalization"  // <-- no comma
  tooltip: "Convert colored (RGB) image to binary with adjustable threshold - Applies a binary threshold..." // <-- wrong tooltip, duplicate key
},
```

**Suggestion:**
```
Remove the stale long `ColorToBinary` tooltip, and add the missing comma after the correct tooltip:

```ts
{
  type: "imageconvertions_histogramequalization",
  message0: "Enhance contrast using histogram equalization",
  previousStatement: null,
  nextStatement: null,
  style: "conversions_style",
  tooltip: "Automatically improves image contrast using histogram equalization",
},
```

Also verify that the `ColorToBinary` block definition still retains its own tooltip (the long descriptive one) in its correct block object, and that the new short tooltip being added to it is intentional.
```

File: imagelab-backend/app/operators/conversions/histogram_equalization.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    """
+
+    def compute(self, image: np.ndarray) -> np.ndarray:
+
+        if image is None:
+            return image
+
+        # Grayscale image
+        if len(image.shape) == 2:
+            return cv2.equalizeHist(image)
```
Comment (ivantha) [inline imagelab-backend/app/operators/conversions/histogram_equalization.py]: **[HIGH]** `cv2.equalizeHist` only accepts `uint8` images — no dtype guard causes runtime crash

`cv2.equalizeHist()` is documented to require an 8-bit (`uint8`) single-channel image. If the pipeline passes a `float32`, `float64`, or `uint16` image (e.g., from a HDR source or after a preceding float-producing operator), OpenCV will throw:

```
cv2.error: (-5:Bad argument) in function 'equalizeHist'
  > Overload resolution failed:
  >  - img data type = 5 is not supported
```

This would crash the entire pipeline with no descriptive error message to the user. The same issue exists for both the grayscale path (line 21) and the colour path (line 29).

**Suggestion:**
```
Add an explicit dtype check (and optionally a clip+cast conversion) before calling `equalizeHist`:

```python
def compute(self, image: np.ndarray) -> np.ndarray:
    if image is None:
        return image

    if image.dtype != np.uint8:
        raise ValueError(
            f"HistogramEqualization requires a uint8 image, got {image.dtype}. "
            "Convert the image to uint8 before applying this operator."
        )

    # ... rest of logic
```

Alternatively, auto-convert if the project conventions allow silent coercion:
```python
if image.dtype != np.uint8:
    image = np.clip(image, 0, 255).astype(np.uint8)
```
```

File: imagelab-backend/app/operators/conversions/histogram_equalization.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            # Convert to YCrCb to preserve color fidelity
+            ycrcb = cv2.cvtColor(image, cv2.COLOR_BGR2YCrCb)
+
+            # Equalize only the luminance channel
+            ycrcb[:, :, 0] = cv2.equalizeHist(ycrcb[:, :, 0])
+
+            # Convert back to BGR
+            return cv2.cvtColor(ycrcb, cv2.COLOR_YCrCb2BGR)
+
+        return image
```
Comment (ivantha) [inline imagelab-backend/app/operators/conversions/histogram_equalization.py]: **[MEDIUM]** 4-channel (BGRA/RGBA) images silently pass through unprocessed

The code handles 2D (grayscale) and 3-channel BGR images, but any image with `image.shape[2] == 4` (e.g., a PNG with an alpha channel) falls through to the final `return image` on line 34 — completely unchanged and with no warning or error. This is a silent no-op that could be very confusing to end users who apply this operator to BGRA images and see no change.

```python
# 4-channel image silently falls here
return image  # line 34 — no equalization applied, no feedback given
```

**Suggestion:**
```
Either handle BGRA images explicitly (equalize Y channel only, preserve alpha), or raise a descriptive error:

```python
# Option A: Handle BGRA by splitting alpha out
if len(image.shape) == 3 and image.shape[2] == 4:
    bgr = image[:, :, :3]
    alpha = image[:, :, 3]
    ycrcb = cv2.cvtColor(bgr, cv2.COLOR_BGR2YCrCb)
    ycrcb[:, :, 0] = cv2.equalizeHist(ycrcb[:, :, 0])
    equalized_bgr = cv2.cvtColor(ycrcb, cv2.COLOR_YCrCb2BGR)
    return np.dstack([equalized_bgr, alpha])

# Option B: Raise early
if len(image.shape) == 3 and image.shape[2] not in (1, 3):
    raise ValueError(f"Unsupported number of channels: {image.shape[2]}")
```
```

Comment (ivantha) [inline imagelab-backend/app/operators/conversions/histogram_equalization.py]: **[MEDIUM]** No automated unit tests added for the new operator

The PR description notes "New tests added: ☐" (unchecked) and only claims manual testing. Given that the operator has multiple code paths (grayscale, colour, None input, unsupported formats), the lack of automated tests means:

- **Regression risk**: A future refactor could silently break histogram equalization without CI catching it.
- **The `float` dtype bug** (see separate finding) would have been caught by a simple parametrised test.
- **Edge cases** like a single-row image, a constant-value image (all pixels identical), and an already-high-contrast image are not tested.

This is especially important since the project already has existing tests that pass.

**Suggestion:**
```
Add unit tests covering at minimum:

```python
# test_histogram_equalization.py
import numpy as np
import pytest
from app.operators.conversions.histogram_equalization import HistogramEqualization

op = HistogramEqualization()

def test_grayscale():
    img = np.zeros((100, 100), dtype=np.uint8)
    img[50:, 50:] = 128
    result = op.compute(img)
    assert result.shape == img.shape
    assert result.dtype == np.uint8

def test_color_bgr():
    img = np.random.randint(0, 256, (100, 100, 3), dtype=np.uint8)
    result = op.compute(img)
    assert result.shape == img.shape
    assert result.dtype == np.uint8

def test_none_input():
    assert op.compute(None) is None

def test_float_image_raises():
    img = np.random.rand(100, 100, 3).astype(np.float32)
    with pytest.raises(ValueError):
        op.compute(img)
```
```

File: imagelab-backend/app/operators/conversions/histogram_equalization.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    """
+    Enhances image contrast using histogram equalization.
+    Works for both grayscale and color images.
+    No parameters required.
+    """
+
+    def compute(self, image: np.ndarray) -> np.ndarray:
+
+        if image is None:
+            return image
```
Comment (ivantha) [inline imagelab-backend/app/operators/conversions/histogram_equalization.py]: **[LOW]** Returning `None` for `None` input silently propagates a null through the pipeline

When `image is None`, the operator silently returns `None`. While this pattern may be consistent with the `BaseOperator` base class convention, it means a null image can flow through the entire pipeline without raising an error, potentially causing a confusing `AttributeError` or `cv2.error` far downstream when the `None` is eventually passed to an operator that doesn't check for it.

It is worth verifying whether the rest of the pipeline handles `None` values explicitly, or whether raising early here would be more appropriate and consistent with how errors are surfaced to the user.

**Suggestion:**
```
Consider raising a clear error or logging a warning:

```python
if image is None:
    raise ValueError("HistogramEqualization received a None image.")
```

If the base class contract explicitly allows returning `None` to signal a no-op, add a comment here explaining that decision so future maintainers understand the intent.
```

**Comment by AbhaBarge** [comment]
@ivantha I’ve rebased the branch on the latest main, resolved merge conflicts, and addressed the backend and frontend lint issues. The workflows appear to be awaiting maintainer approval. Please approve.
Thank you !

**Comment by AbhaBarge** [comment]
@ivantha Please check. All the issues have been resolved now. I apologise for the immense delay. Thank you for your patience, I learnt a lot while working on this single PR !
image

**Comment by ivantha** [comment]
Reviewed. Adds a HistogramEqualization operator (grayscale via equalizeHist, color via Y-channel in YCrCb) wired through the backend registry and a Blockly block in the Conversions category. PR also includes some incidental changes (kernelSize validation on morphological, formatting tweaks in color.py, shortened colortobinary tooltip) which overlap slightly with other in-flight PRs but are benign. Merging via rebase.

**Comment by ivantha** [comment]
Attempted rebase-merge but GitHub reports the branch can't be rebased cleanly (likely because of non-linear history or a merge commit on the PR branch). Could you rebase this branch onto the latest main locally and force-push? I'll re-attempt the merge afterward.
